README.md
# SiriDB Connector using libuv and libsiridb.
This example is written as an extension to libsiridb which we will call libsuv.

---------------------------------------
  * [API](#api)
    * [suv_buf_t](#suv_buf_t)
    * [suv_write_t](#suv_write_t)
    * [suv_connect_t](#suv_connect_t)
    * [suv_query_t](#suv_query_t)
    * [suv_insert_t](#suv_insert_t)
    * [Miscellaneous functions](#miscellaneous-functions)

---------------------------------------

## API
>Note: libsuv uses some of the public data members which are exposed by libsiridb.
>In most cases libsuv exposes its own public space for user-defined data which
>should be used instead.

### `suv_buf_t`
Buffer type. Each SiriDB connection must have its own buffer.

#### `suv_buf_t * suv_buf_create(siridb_t * siridb)`
Create and return a new buffer. A `siridb_t` instance is required.

>Warning: Do not use `siridb->data` since it will be overwritten as soon as the
>buffer is used as a connection.

*Public members*
- `void * suv_buf_t.data`: Space for user-defined arbitrary data. libsuv does
not use this field.
- `suv_cb onclose`: Can be set to an optional callback function which will be
called when a connection is closed.
- `suv_cb onerror`: Can be set to an optional callback function which will be
called when the normal request callback (`siridb_on_pkg()`) returns with an error.

#### `suv_buf_from_req(siridb_req_t * req)`
Macro function to get the `suv_buf_t*` from a request.

#### `void suv_buf_destroy(suv_buf_t * suvbf)`
Cleanup a buffer. Call this function after the connection is closed.

#### `void suv_close(suv_buf_t * buf, const char * msg)`
Close a connection. As long as the buffer is not destroyed, the same `suv_buf_t`
can be used again to create a new connection. Argument `msg` is allowed to be
`NULL` in wich case the default message will be used. The message (or default message)
will be parsed to the `suv_buf_t.onclose` callback function.

### `suv_write_t`
General write type. Used to communicate with SiriDB.

*Public members*
- `void * suv_write_t.data`: Space for user-defined arbitrary data. libsuv does
not use this field.
- `void * suv_write_t.pkg`: Contains the package to send. (readonly)

#### `void suv_write_destroy(suv_connect_t * connect)`
Cleanup a write handle. This function should be called from a request
(`siridb_req_t`) callback function.

#### `void suv_write_error(suv_write_t * swrite, int err_code)`
Used to cancel a write handle. This will set the request status to `err_code`,
removes the request from the queue and calls the request callback function.

### `suv_connect_t`
Connect handle. Alias for `suv_write_t`.

#### `suv_connect_t * suv_connect_create(siridb_req_t * req, const char * username, const char * password, const char * dbname)`
Create and return a connection handle. After the connection handle is created,
you must manually bind the handle to `req->data`. This must be done explicit to
make clear that you are also responsible for handling the cleanup.

Returns `NULL` in case of a memory allocation error.

#### `void suv_connect_destroy(suv_connect_t * connect)`
Cleaunp a connection handle. This function should be called from a request
(`siridb_req_t`) callback function. Alias for `suv_write_destroy()`.

#### `void suv_connect(uv_loop_t * loop, suv_connect_t * connect, suv_buf_t * buf, struct sockaddr * addr)`
Connect and authenticate to SiriDB.

>Warning: This function overwrites the members `buf->siridb->data` so
>you should not use this property. Instead the public members `buf->data` and
>`connect->data` are available and safe to use.

Example:
```c
#include <suv.h>

uv_loop_t loop;

void connect_cb(siridb_req_t * req)
{
    suv_connect_t * connect = (suv_connect_t *) req->data;

    if (req->status) {
        printf("connect or auth failed: %s\n", siridb_strerror(req->status));
    } else if (req->pkg->tp != CprotoResAuthSuccess) {
        printf("authentication failed (error %u)\n", req->pkg->tp);
    } else {
        // do something with the connection
    }

    /* cleanup connetion handle */
    suv_connect_destroy(connect);

    /* lets stop the example */
    suv_close(suv_buf_from_req(req), NULL);

    /* cleanup connection request */
    siridb_req_destroy(req);
}

int main(void)
{
    struct sockaddr_in addr;
    /* initialize uv loop */
    uv_loop_init(&loop);

    /* asume siridb-server is running on localhost and port 9000 */
    uv_ip4_addr("127.0.0.1", 9000, &addr);

    /* create a siridb client */
    siridb_t * siridb = siridb_create();

    /* create a buffer for the connection */
    suv_buf_t * buf = suv_buf_create(siridb);

    /* create a connection request */
    siridb_req_t * req = siridb_req_create(siridb, connect_cb, NULL);

    /* create a connection handle */
    suv_connect_t * connect = suv_connect_create(req, "iris", "siri", "dbtest");

    /* explicit bind the connect handle to the request. (this must be done!) */
    req->data = (void *) connect;

    /* Warning: This overwrites tcp->data and siridb->data so do not use these
     *          members yourself. */
    suv_connect(&loop, connect, buf, (struct sockaddr *) &addr);

    /* run the uv event loop */
    uv_run(&loop, UV_RUN_DEFAULT);

    /* close the loop */
    uv_loop_close(&loop);

    /* cleanup buffer */
    suv_buf_destroy(buf);

    /* cleanup siridb */
    siridb_destroy(siridb);

    return 0;
}
```

### `suv_query_t`
Query handle. Alias for `suv_write_t`.

#### `suv_query_t * suv_query_create(siridb_req_t * req, const char * query)`
Create and return a query handle. After the query handle is created,
you must manually bind the handle to `req->data`. This must be done explicit to
make clear that you are also responsible for handling the cleanup.

Returns `NULL` in case of a memory allocation error.

#### `void suv_query_destroy(suv_query_t * suvq)`
Cleaunp a query handle. This function should be called from a request
(`siridb_req_t`) callback function. Alias for `suv_write_destroy()`.

#### `void suv_query(suv_query_t * suvq)`
Query SiriDB.

Example
```c
/* first create a request */
siridb_req_t * req = siridb_req_create(siridb, example_cb, NULL);

/* create query handle */
suv_query_t * handle = suv_query_create(req, "select * from 'my-series'");

/* bind query handle to req->data */
req->data = (void *) handle;

/* now run the query, check the callback for the result or errors. */
suv_query(handle);
```

Example callback function:
```c
void example_cb(siridb_req_t * req)
{
    if (req->status != 0) {
        printf("error handling request: %s", siridb_strerror(req->status));
    } else {
        /* get the response */
        siridb_resp_t * resp = siridb_resp_create(req->pkg, NULL);

        // do something with the response...

        // a general cb function could do something based on the response type...
        switch(resp->tp) {
        case SIRIDB_RESP_TP_UNDEF:
        case SIRIDB_RESP_TP_SELECT:
        case SIRIDB_RESP_TP_LIST:
        case SIRIDB_RESP_TP_SHOW:
        case SIRIDB_RESP_TP_COUNT:
        case SIRIDB_RESP_TP_CALC:
        case SIRIDB_RESP_TP_SUCCESS:
        case SIRIDB_RESP_TP_SUCCESS_MSG:
        case SIRIDB_RESP_TP_ERROR:
        case SIRIDB_RESP_TP_ERROR_MSG:
        case SIRIDB_RESP_TP_HELP:
        case SIRIDB_RESP_TP_MOTD:
        case SIRIDB_RESP_TP_DATA: break;
        }

        /* cleanup response */
        siridb_resp_destroy(resp);
    }

    /* destroy handle */
    suv_write_destroy((suv_write_t *) req->data);

    /* destroy request */
    siridb_req_destroy(req);
}
```

### `suv_insert_t`
Insert handle. Alias for `suv_write_t`.

#### `suv_insert_t * suv_insert_create(siridb_req_t * req, siridb_series_t * series[], size_t n)`
Create and return an insert handle. Argument `n` must be equal or smaller than
the number of series in the `series[]` array.

After the insert handle is created, you must manually bind the handle to
`req->data`. This must be done explicit to make clear that you are also
responsible for handling the cleanup.

Returns `NULL` in case of a memory allocation error.

#### `void suv_insert_destroy(suv_insert_t * insert)`
Cleanup an insert handle. This function should be called from a request
(`siridb_req_t`) callback function. Alias for `suv_write_destroy()`.

#### `void suv_insert(suv_insert_t * insert`
Insert data into SiriDB.

Example:
```c
/* create an array of series, just one for this example */
siridb_series_t * series[1];

/* create series */
series[0] = siridb_series_create(
    SIRIDB_SERIES_TP_INT64,         /* type int64, could also be double */
    "example-series-name",          /* some name for the series */
    10);                            /* number of points */

/* create some sample points */
for (size_t i = 0; i < series[0]->n; i++) {
    siridb_point_t * point = series[0]->points + i;
    point->ts = (uint64_t) time(NULL) - series[0]->n + i;
    point->via.int64 = (int64_t) i;
}

/* create a request */
siridb_req_t * req = siridb_req_create(siridb, example_cb, NULL);

/* create insert handle */
suv_insert_t * handle = suv_insert_create(req, series, 1);

/* bind the handle to the request */
req->data = (void *) handle;

/* cleanup the series since the series is now packed in handle->pkg */
siridb_series_destroy(series[0]);

/* insert data into siridb, the callback should be checked for errors.
 * (a successful insert has a SIRIDB_RESP_TP_SUCCESS_MSG siridb_resp_t.tp) */
suv_insert(handle);
```

### Miscellaneous functions
#### `const char * suv_strerror(int err_code)`
Returns the error message for a given error code.

#### `const char * suv_version(void)`
Returns the version of libsuv.


ChangeLog
libsuv (0.1.2)

  * Fixed bug in cleanup.

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  11 Sep 2017

libsuv (0.1.1)

  * Cleanup connection when connect has failed.

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  11 Sep 2017

libsuv (0.1.0)

  * Initial version.

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  31 Aug 2017

suv.c
/*
 * suv.c - SiriDB C-Connector libuv
 *
 *  Created on: Aug 31, 2017
 *      Author: Jeroen van der Heijden <jeroen@transceptor.technology>
 */

#include "suv.h"
#include <string.h>
#include <assert.h>

static suv_write_t * suv__write_create(void);
static void suv__close_tcp(uv_handle_t * tcp);
static void suv__write(suv_write_t * swrite);
static void suv__alloc_buf(uv_handle_t * handle, size_t sugsz, uv_buf_t * buf);
static void suv__on_data(uv_stream_t * clnt, ssize_t n, const uv_buf_t * buf);
static void suv__write_cb(uv_write_t * uvreq, int status);
static void suv__connect_cb(uv_connect_t * uvreq, int status);

const long int MAX_PKG_SIZE = 209715200; // can be changed to anything you want

/*
 * Return libsuv version info.
 */
const char * suv_version(void)
{
    return SUV_VERSION;
}

/*
 * Create and returns a buffer object or NULL in case of an allocation error.
 */
suv_buf_t * suv_buf_create(siridb_t * siridb)
{
    suv_buf_t * suvbf = (suv_buf_t *) malloc(sizeof(suv_buf_t));
    if (suvbf != NULL)
    {
        suvbf->siridb = siridb;
        suvbf->len = 0;
        suvbf->size = 0;
        suvbf->buf = NULL;
        suvbf->onclose = NULL;
        suvbf->onerror = NULL;
    }
    return suvbf;
}

/*
 * Destroy a buffer object.
 */
void suv_buf_destroy(suv_buf_t * suvbf)
{
    uv_tcp_t * tcp_ = (uv_tcp_t *) suvbf->siridb->data;
    if (tcp_ != NULL)
    {
        suv_close(suvbf, NULL);
    }
    free(suvbf->buf);
    free(suvbf);
}

/*
 * Create and return an connect object or NULL in case of an allocation error.
 */
suv_connect_t * suv_connect_create(
    siridb_req_t * req,
    const char * username,
    const char * password,
    const char * dbname)
{
    assert (req->data == NULL); /* req->data should be set to -this- */

    suv_write_t * connect = suv__write_create();
    if (connect != NULL)
    {
        connect->pkg = siridb_pkg_auth(req->pid, username, password, dbname);
        connect->_req = req;
        if (connect->pkg == NULL)
        {
            suv_write_destroy(connect);
            connect = NULL;
        }
    }
    return (suv_connect_t *) connect;
}

/*
 * Destroy an connect object.
 */
void suv_connect_destroy(suv_connect_t * connect)
{
    suv_write_destroy((suv_write_t * ) connect);
}

/*
 * Use this function to connect to SiriDB. Always use the callback defined by
 * the request object parsed to suv_connect_create() for errors.
 */
void suv_connect(
    uv_loop_t * loop,
    suv_connect_t * connect,
    suv_buf_t * buf,
    struct sockaddr * addr)
{
    assert (connect->_req->data == connect);  /* bind connect to req->data */

    uv_connect_t * uvreq = (uv_connect_t *) malloc(sizeof(uv_connect_t));
    if (uvreq == NULL)
    {
        suv_write_error((suv_write_t *) connect, ERR_MEM_ALLOC);
        return;
    }

    uv_tcp_t * tcp_ = (uv_tcp_t *) malloc(sizeof(uv_tcp_t));
    if (tcp_ == NULL)
    {
        suv_write_error((suv_write_t *) connect, ERR_MEM_ALLOC);
        return;
    }

    tcp_->data = (void *) buf;
    buf->siridb->data = (void *) tcp_;

    uv_tcp_init(loop, tcp_);

    uvreq->data = (void *) connect->_req;
    uv_tcp_connect(uvreq, tcp_, addr, suv__connect_cb);
}

/*
 * Close an open connection.
 */
void suv_close(suv_buf_t * buf, const char * msg)
{
    uv_tcp_t * tcp_ = (uv_tcp_t *) buf->siridb->data;
    if (tcp_ != NULL && !uv_is_closing((uv_handle_t *) tcp_))
    {
        if (buf->onclose != NULL)
        {
            buf->onclose(buf->data, (msg == NULL) ? "connection closed" : msg);
        }
        uv_close((uv_handle_t *) tcp_, suv__close_tcp);
    }
}

/*
 * Create and return a query object or NULL in case of an allocation error.
 */
suv_query_t * suv_query_create(siridb_req_t * req, const char * query)
{
    assert (req->data == NULL); /* req->data should be set to -this- */

    suv_write_t * suvq = suv__write_create();
    if (suvq != NULL)
    {
        suvq->pkg = siridb_pkg_query(req->pid, query);
        suvq->_req = req;
        if (suvq->pkg == NULL)
        {
            suv_write_destroy(suvq);
            suvq = NULL;
        }
    }
    return (suv_query_t *) suvq;
}

/*
 * Destroy a query object.
 */
void suv_query_destroy(suv_query_t * suvq)
{
    suv_write_destroy((suv_write_t * ) suvq);
}

/*
 * This function actually runs a query. Always use the callback defined by
 * the request object parsed to suv_query_create() for errors.
 */
void suv_query(suv_query_t * suvq)
{
    suv__write((suv_write_t *) suvq);
}

/*
 * Create and return a insert object or NULL in case of an allocation error.
 */
suv_insert_t * suv_insert_create(
    siridb_req_t * req,
    siridb_series_t * series[],
    size_t n)
{
    assert (req->data == NULL); /* req->data should be set to -this- */

    suv_write_t * insert = suv__write_create();
    if (insert != NULL)
    {
        insert->pkg = siridb_pkg_series(req->pid, series, n);
        insert->_req = req;
        if (insert->pkg == NULL)
        {
            suv_write_destroy(insert);
            insert = NULL;
        }
    }
    return (suv_insert_t *) insert;
}

/*
 * Destroy a insert object.
 */
void suv_insert_destroy(suv_insert_t * insert)
{
    suv_write_destroy((suv_write_t * ) insert);
}

/*
 * This function actually send the insert. Always use the callback defined by
 * the request object parsed to suv_insert_create() for errors.
 */
void suv_insert(suv_insert_t * insert)
{
    suv__write((suv_write_t *) insert);
}

/*
 * Create and return a write object or NULL in case of an allocation error.
 */
static suv_write_t * suv__write_create(void)
{
    suv_write_t * swrite = (suv_write_t *) malloc(sizeof(suv_write_t));
    if (swrite != NULL)
    {
        swrite->data = NULL;
        swrite->pkg = NULL;
    }
    return swrite;
}

/*
 * Close and free handle.
 */
static void suv__close_tcp(uv_handle_t * tcp)
{
    if (tcp->data != NULL)
    {
        suv_buf_t * buf = (suv_buf_t *) tcp->data;
        if (tcp == (uv_handle_t *) buf->siridb->data)
        {
            buf->siridb->data = NULL;
        }
    }
    free(tcp);
}

/*
 * Destroy a write object.
 */
void suv_write_destroy(suv_write_t * swrite)
{
    free(swrite->pkg);
    free(swrite);
}

/*
 * Set request error and run callback
 */
void suv_write_error(suv_write_t * swrite, int err_code)
{
    queue_pop(swrite->_req->siridb->queue, swrite->pkg->pid);
    swrite->_req->status = err_code;
    swrite->_req->cb(swrite->_req);
}

/*
 * Return error string. UV errors should be set to positive values.
 */
const char * suv_strerror(int err_code)
{
    return (err_code > 0) ?
            uv_strerror(-err_code) : siridb_strerror(err_code);
}

/*
 * This function actually send the data.
 */
static void suv__write(suv_write_t * swrite)
{
    assert (swrite->_req->data == swrite); /* bind swrite to req->data */

    uv_stream_t * stream = (uv_stream_t *) swrite->_req->siridb->data;
    if (stream == NULL)
    {
        suv_write_error(swrite, ERR_SOCK_WRITE);
        return;
    }

    uv_write_t * uvreq = (uv_write_t *) malloc(sizeof(uv_write_t));
    if (uvreq == NULL)
    {
        suv_write_error(swrite, ERR_MEM_ALLOC);
        return;
    }

    uvreq->data = (void *) swrite->_req;

    uv_buf_t buf = uv_buf_init(
        (char *) swrite->pkg,
        sizeof(siridb_pkg_t) + swrite->pkg->len);

    uv_write(uvreq, stream, &buf, 1, suv__write_cb);
}

static void suv__connect_cb(uv_connect_t * uvreq, int status)
{
    siridb_req_t * req = (siridb_req_t *) uvreq->data;
    suv_connect_t * connect = (suv_connect_t *) req->data;

    if (status != 0)
    {
        /* error handling */
        suv_write_error((suv_write_t *) connect, -status);
        uv_close((uv_handle_t *) connect->_req->siridb->data, suv__close_tcp);
    }
    else
    {
        uv_write_t * uvw = (uv_write_t *) malloc(sizeof(uv_write_t));
        if (uvw == NULL)
        {
            suv_write_error((suv_write_t *) connect, ERR_MEM_ALLOC);
            uv_close(
                (uv_handle_t *) connect->_req->siridb->data, suv__close_tcp);
        }
        else
        {
            uvw->data = (void *) req;

            uv_buf_t buf = uv_buf_init(
                    (char *) connect->pkg,
                    sizeof(siridb_pkg_t) + connect->pkg->len);

            uv_read_start(uvreq->handle, suv__alloc_buf, suv__on_data);
            uv_write(uvw, uvreq->handle, &buf, 1, suv__write_cb);
        }
    }
    free(uvreq);
}

static void suv__alloc_buf(uv_handle_t * handle, size_t sugsz, uv_buf_t * buf)
{
    suv_buf_t * suvbf = (suv_buf_t *) handle->data;
    if (suvbf->len == 0 && suvbf->size != sugsz)
    {
        free(suvbf->buf);
        suvbf->buf = (char *) malloc(sugsz);
        if (suvbf->buf == NULL)
        {
            abort(); /* memory allocation error */
        }
        suvbf->size = sugsz;
        suvbf->len = 0;
    }

    buf->base = suvbf->buf + suvbf->len;
    buf->len = suvbf->size - suvbf->len;
}

static void suv__on_data(uv_stream_t * clnt, ssize_t n, const uv_buf_t * buf)
{
    suv_buf_t * suvbf = (suv_buf_t *) clnt->data;
    siridb_pkg_t * pkg;
    size_t total_sz;
    int rc;

    if (n < 0)
    {
        suv_close(suvbf, (n != UV_EOF) ? uv_strerror(n) : NULL);
        return;
    }

    suvbf->len += n;

    if (suvbf->len < sizeof(siridb_pkg_t))
    {
        return;
    }

    pkg = (siridb_pkg_t *) suvbf->buf;
    if (!siridb_pkg_check_bit(pkg) || pkg->len > MAX_PKG_SIZE)
    {
        suv_close(suvbf, "invalid package, connection closed");
        return;
    }

    total_sz = sizeof(siridb_pkg_t) + pkg->len;

    if (suvbf->len < total_sz)
    {
        if (suvbf->size < total_sz)
        {
            char * tmp = realloc(suvbf->buf, total_sz);
            if (tmp == NULL)
            {
                abort(); /* memory allocation error */
            }
            suvbf->buf = tmp;
            suvbf->size = total_sz;
        }
        return;
    }

    if ((rc = siridb_on_pkg(suvbf->siridb, pkg)))
    {
        if (suvbf->onerror != NULL)
        {
            suvbf->onerror(suvbf->data, siridb_strerror(rc));
        }
    }

    suvbf->len -= total_sz;

    if (suvbf->len > 0)
    {
        /* move data and call suv_on_data() function again */
        memmove(suvbf->buf, suvbf->buf + total_sz, suvbf->len);
        suv__on_data(clnt, 0, buf);
    }
}

static void suv__write_cb(uv_write_t * uvreq, int status)
{
    if (status)
    {
        /* error handling */
        siridb_req_t * req = (siridb_req_t *) uvreq->data;
        suv_write_t * swrite = (suv_write_t *) req->data;

        suv_write_error(swrite, -status);
    }

    /* free uv_write_t */
    free(uvreq);
}




suv.h
/*
 * suv.h - SiriDB C-Connector, example using libuv
 *
 *  Created on: Jun 09, 2017
 *      Author: Jeroen van der Heijden <jeroen@transceptor.technology>
 */

#ifndef SUV_H_
#define SUV_H_

#define SUV_VERSION_MAJOR 0
#define SUV_VERSION_MINOR 1
#define SUV_VERSION_PATCH 2

#define SUV_STRINGIFY(num) #num
#define SUV_VERSION_STR(major,minor,patch)   \
    SUV_STRINGIFY(major) "."                 \
    SUV_STRINGIFY(minor) "."                 \
    SUV_STRINGIFY(patch)

#define SUV_VERSION SUV_VERSION_STR(          \
        SUV_VERSION_MAJOR,                   \
        SUV_VERSION_MINOR,                   \
        SUV_VERSION_PATCH)

#include <stdlib.h>
#include <uv.h>
#include <libsiridb/siridb.h>

/* type definitions */
typedef struct suv_buf_s suv_buf_t;
typedef struct suv_write_s suv_write_t;
typedef struct suv_write_s suv_connect_t;
typedef struct suv_write_s suv_query_t;
typedef struct suv_write_s suv_insert_t;

/* public functions */
#ifdef __cplusplus
extern "C" {
#endif

typedef void (*suv_cb) (void * buf_data, const char * msg);

suv_buf_t * suv_buf_create(siridb_t * siridb);
void suv_buf_destroy(suv_buf_t * suvbf);

void suv_write_destroy(suv_write_t * swrite);
void suv_write_error(suv_write_t * swrite, int err_code);

suv_connect_t * suv_connect_create(
    siridb_req_t * req,
    const char * username,
    const char * password,
    const char * dbname);
void suv_connect_destroy(suv_connect_t * connect);
void suv_connect(
    uv_loop_t * loop,
    suv_connect_t * connect,
    suv_buf_t * buf,
    struct sockaddr * addr);
void suv_close(suv_buf_t * buf, const char * msg);

suv_query_t * suv_query_create(siridb_req_t * req, const char * query);
void suv_query_destroy(suv_query_t * suvq);
void suv_query(suv_query_t * suvq);

suv_insert_t * suv_insert_create(
    siridb_req_t * req,
    siridb_series_t * series[],
    size_t n);
void suv_insert_destroy(suv_insert_t * insert);
void suv_insert(suv_insert_t * insert);

const char * suv_strerror(int err_code);
const char * suv_version(void);

#define suv_buf_from_req(REQ__) \
    ((suv_buf_t *) ((uv_tcp_t *) REQ__->siridb->data)->data)

#ifdef __cplusplus
}
#endif

/* struct definitions */
struct suv_buf_s
{
    void * data;            /* public */
    suv_cb onclose;         /* public */
    suv_cb onerror;         /* public */
    char * buf;
    size_t len;
    size_t size;
    siridb_t * siridb;
};

struct suv_write_s
{
    void * data;            /* public */
    siridb_pkg_t * pkg;     /* packge to send */
    siridb_req_t * _req;    /* will not be cleared */
};

#endif /* SUV_H_ */


example/main.c
/*
 * main.c
 *    SiriDB C-Connector example using libsuv. This is a quite large example
 *    since we try to cover most features from libsiridb and libsuv.
 *
 *  WARNING:
 *     Be carefull runing this example since it will query and insert data into
 *     SiriDB!
 *
 *  Compile using:
 *
 *     gcc main.c -lsuv -lsiridb -lqpack -luv -o example.out
 *
 *  Created on: Aug 31, 2017
 *      Author: Jeroen van der Heijden <jeroen@transceptor.technology>
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <suv.h>

static uv_loop_t loop;

/* Change this values to your needs */
const char * SERVER     = "127.0.0.1";
const int    PORT       = 9000;
const char * USER       = "iris";
const char * PASSWD     = "siri";
const char * DBNAME     = "dbtest";
const char * QUERY      = "show";

static void connect_cb(siridb_req_t * req);
static void query_cb(siridb_req_t * req);
static void insert_cb(siridb_req_t * req);
static void insert_example(siridb_t * siridb);
static void send_example_query(siridb_t * siridb, const char * query);
static void print_resp(siridb_resp_t * resp);
static void print_timeit(siridb_timeit_t * timeit);
static void print_select(siridb_select_t * select);
static void print_list(siridb_list_t * list);
static void print_count(uint64_t count);
static void print_calc(uint64_t calc);
static void print_show(siridb_show_t * show);
static void print_msg(const char * msg);
static void on_close(void * data, const char * msg);
static void on_error(void * data, const char * msg);

int main(void)
{
    printf("Running example using:\n"
           " - libqpack %s\n"
           " - libsiridb %s\n"
           " - libsuv %s\n"
           "\n", qp_version(), siridb_version(), suv_version());

    struct sockaddr_in addr;
    char * query = strdup(QUERY);

    uv_loop_init(&loop);

    uv_ip4_addr(SERVER, PORT, &addr);

    siridb_t * siridb = siridb_create();
    /* handle siridb == NULL */
    /* Warning: do not use siridb->data since it will be used by libsuv */

    suv_buf_t * buf = suv_buf_create(siridb);
    /* handle buf == NULL */

    /* set optional callback functions */
    buf->onclose = on_close;
    buf->onerror = on_error;

    siridb_req_t * req = siridb_req_create(siridb, connect_cb, NULL);
    /* handle req == NULL */

    suv_connect_t * connect = suv_connect_create(req, USER, PASSWD, DBNAME);
    /* handle connect == NULL */

    /* for the example we bind a query string but this could be anything */
    connect->data = (void *) query;
    req->data = (void *) connect;

    suv_connect(&loop, connect, buf, (struct sockaddr *) &addr);

    uv_run(&loop, UV_RUN_DEFAULT);

    /* cleanup buffer */
    suv_buf_destroy(buf);

    /* cleanup siridb */
    siridb_destroy(siridb);

    /* close uv loop */
    uv_loop_close(&loop);

    return 0;
}

static void connect_cb(siridb_req_t * req)
{
    suv_connect_t * connect = (suv_connect_t *) req->data;
    /* handle connect == NULL (for example in case the request is cancelled
     * before a connection handle was attached) */

    char * query = (char *) connect->data;
    if (req->status)
    {
        printf("connect failed: %s\n", suv_strerror(req->status));
    }
    else
    {
        switch (req->pkg->tp)
        {
        case CprotoResAuthSuccess:
            send_example_query(req->siridb, query);
            break;
        case CprotoErrAuthCredentials:
            printf("auth failed: invalid credentials\n");
            break;
        case CprotoErrAuthUnknownDb:
            printf("auth failed: unknown database\n");
            break;
        default:
            printf("auth failed: unknown error (%u)\n", req->pkg->tp);
        }
    }

    /* free query string */
    free(query);

    /* destroy suv_connect_t */
    suv_connect_destroy(connect);

    /* destroy siridb request */
    siridb_req_destroy(req);
}

static void query_cb(siridb_req_t * req)
{
    siridb_t * siridb = req->siridb;

    if (req->status != 0)
    {
        printf("error handling request: %s", suv_strerror(req->status));
    }
    else
    {
        /* We can get the output as a JSON string... */
        char * json = qp_sprint(req->pkg->data, req->pkg->len);
        printf("Response as JSON:\n%s\n\n", json);
        free(json);

        /* ..or get a nice response object */
        siridb_resp_t * resp = siridb_resp_create(req->pkg, NULL);
        /* handle resp == NULL or use rc code for details */
        print_resp(resp);

        siridb_resp_destroy(resp);
    }

    /* destroy suv_query_t */
    suv_query_destroy((suv_query_t *) req->data);

    /* destroy siridb request */
    siridb_req_destroy(req);

    /* call the insert example */
    insert_example(siridb);
}

static void insert_cb(siridb_req_t * req)
{
    if (req->status != 0)
    {
        printf("error handling request: %s", suv_strerror(req->status));
    }
    else
    {
        siridb_resp_t * resp = siridb_resp_create(req->pkg, NULL);
        /* handle resp == NULL or use rc code for details */
        print_resp(resp);

        siridb_resp_destroy(resp);
    }

    /* destroy suv_insert_t */
    suv_insert_destroy((suv_insert_t *) req->data);

    /* here we quit the example by closing the handle */
    suv_close(suv_buf_from_req(req), NULL);

    /* destroy siridb request */
    siridb_req_destroy(req);
}

static void insert_example(siridb_t * siridb)
{
    siridb_series_t * series[2]; /* in this example we insert 2 series */
    siridb_point_t * point;

    series[0] = siridb_series_create(
        SIRIDB_SERIES_TP_INT64,         /* type integer */
        "c-conn-int64-test-series",     /* some name for the series */
        10);                            /* number of points */

    for (size_t i = 0; i < series[0]->n; i++)
    {
        point = series[0]->points + i;
        /* set the time-stamp */
        point->ts = (uint64_t) time(NULL) - series[0]->n + i;
        /* set a value, just the values 0 to 9 in this example */
        point->via.int64 = (int64_t) i;
    }

    series[1] = siridb_series_create(
        SIRIDB_SERIES_TP_REAL,          /* type float */
        "c-conn-real-test-series",      /* some name for the series */
        5);                             /* number of points */

    for (size_t i = 0; i < series[1]->n; i++)
    {
        point = series[1]->points + i;
        /* set the time-stamp */
        point->ts = (uint64_t) time(NULL) - series[1]->n + i;
        /* set a value, just the values 0.0 to 0.4 in this example */
        point->via.real = (double) i / 10;
    }

    siridb_req_t * req = siridb_req_create(siridb, insert_cb, NULL);
    /* handle req == NULL */

    suv_insert_t * suvinsert = suv_insert_create(req, series, 2);
    /* handle suvinsert == NULL */

    /* cleanup the series, we don't need them anymore */
    siridb_series_destroy(series[0]);
    siridb_series_destroy(series[1]);

    /* bind suvinsert to qreq->data */
    req->data = (void *) suvinsert;

    suv_insert(suvinsert);
    /* check insert_cb for errors */
}

static void send_example_query(siridb_t * siridb, const char * query)
{
    siridb_req_t * req = siridb_req_create(siridb, query_cb, NULL);
    /* handle req == NULL */

    suv_query_t * suvquery = suv_query_create(req, query);
    /* handle suvquery == NULL */

    /* bind suvquery to req->data */
    req->data = (void *) suvquery;

    suv_query(suvquery);
    /* check query_cb for errors */
}

static void on_close(void * data, const char * msg)
{
    printf("%s\n", msg);
}

static void on_error(void * data, const char * msg)
{
    printf("got an error: %s\n", msg);
}

static void print_resp(siridb_resp_t * resp)
{
    print_timeit(resp->timeit);

    switch (resp->tp)
    {
    case SIRIDB_RESP_TP_SELECT:
        print_select(resp->via.select); break;
    case SIRIDB_RESP_TP_LIST:
        print_list(resp->via.list); break;
    case SIRIDB_RESP_TP_COUNT:
        print_count(resp->via.count); break;
    case SIRIDB_RESP_TP_CALC:
        print_calc(resp->via.calc); break;
    case SIRIDB_RESP_TP_SHOW:
        print_show(resp->via.show); break;
    case SIRIDB_RESP_TP_SUCCESS:
        print_msg(resp->via.success); break;
    case SIRIDB_RESP_TP_SUCCESS_MSG:
        print_msg(resp->via.success_msg); break;
    case SIRIDB_RESP_TP_ERROR:
        print_msg(resp->via.error); break;
    case SIRIDB_RESP_TP_ERROR_MSG:
        print_msg(resp->via.error_msg); break;
    default: assert(0);
    }
}

static void print_timeit(siridb_timeit_t * timeit)
{
    if (timeit != NULL)
    {
        printf("Query time: %f seconds\n", timeit->perfs[timeit->n - 1].time);
        for (size_t i = 0; i < timeit->n; i++)
        {
            printf(
                "    server: %s time: %f\n",
                timeit->perfs[i].server,
                timeit->perfs[i].time);
        }
        printf("\n");
    }
}

static void print_select(siridb_select_t * select)
{
    printf("Select response for %" PRIu64 " series:\n", select->n);
    for (size_t m = 0; m < select->n; m++)
    {
        siridb_series_t * series = select->series[m];

        printf("    series: '%s'\n", series->name);
        for (size_t i = 0; i < series->n; i++)
        {
            printf(
                "        timestamp: %" PRIu64 " value: ",
                series->points[i].ts);
            switch (series->tp)
            {
            case SIRIDB_SERIES_TP_INT64:
                printf("%ld\n", series->points[i].via.int64); break;
            case SIRIDB_SERIES_TP_REAL:
                printf("%f\n", series->points[i].via.real); break;
            case SIRIDB_SERIES_TP_STR:
                printf("%s\n", series->points[i].via.str); break;
            }
        }
    }
}

static void print_list(siridb_list_t * list)
{
    printf(
        "List response with %zu columns and %zu rows:\n",
        list->headers->via.array->n,
        list->data->via.array->n);
    for (size_t r = 0; r < list->data->via.array->n; r++)
    {
        qp_array_t * row = list->data->via.array->values[r].via.array;
        for (size_t c = 0; c < row->n; c++)
        {
            if (c) printf(", ");
            qp_res_fprint(row->values + c, stdout);
        }
        printf("\n");
    }
}

static void print_count(uint64_t count)
{
    printf("Count response: %" PRIu64 "\n", count);
}

static void print_calc(uint64_t calc)
{
    printf("Calc response: %lu\n", calc);
}

static void print_show(siridb_show_t * show)
{
    printf("Show response with %zu items\n", show->n);
    for (size_t i = 0; i < show->n; i++)
    {
        printf("    %s: ", show->items[i].key);
        qp_res_fprint(show->items[i].value, stdout);
        printf("\n");
    }
}

static void print_msg(const char * msg)
{
    printf("%s\n", msg);
}
